### PR TITLE
Send Page amount USD.

### PR DESCRIPTION
**Pull request by Sirmorrison**
### Resolves

Issue #142 

### What's new
This PR adds the send amount USD equivalence for all transactions fields to the send Page.

### Relevant screenshots or logs
image

image

image

### CODE REVIEW

**Comment by oshorefueled** [comment]
The amount-usd feature doesn't work. The currency conversion feature isn't just to show a user DCR equivalent in dollars but to also allow the user input an amount in dollars and have its DCR equivalent sent. Currently, it doesn't do that.

**Comment by oshorefueled** [comment]
The position of the conversion button and values look off, we should be able to make it look just like the mobile. Kindly have a look at the related issue for the mobile mockup. 

Screenshot 2020-06-19 at 7 24 41 AM

**Comment by Sirmorrison** [comment]
> The position of the conversion button and values look off, we should be able to make it look just like the mobile. Kindly have a look at the related issue for the mobile mockup.
> 
> Screenshot 2020-06-19 at 7 24 41 AM

Resolved.

ezgif com-video-to-gif(9)

File: ui/send_page.go [hunk 1]
```diff
@@ -16,6 +19,10 @@ import (
 	"golang.org/x/exp/shiny/materialdesign/icons"
 )
 
+type data struct {
```
Comment (oshorefueled) [inline ui/send_page.go]: why do we need a new data type to handle just one property?

File: ui/send_page.go [hunk 2]
```diff
@@ -260,6 +305,9 @@ func (pg *SendPage) setAccountTabs() {
 }
 
 func (pg *SendPage) Layout(common pageCommon) {
+	go pg.getUSDValues(&pg.data)
```
Comment (oshorefueled) [inline ui/send_page.go]: This method is going to be called repeatedly, and even though it's called from a go routine for optimal reasons that's still a lot of go routines to spin up. Wouldn't it be better if this was only called when a user types the amount into the amount field?

File: ui/send_page.go [hunk 3]
```diff
@@ -400,13 +448,13 @@ func (pg *SendPage) drawSelectedAccountSection(gtx *layout.Context) {
 func (pg *SendPage) drawTransactionDetailWidgets(gtx *layout.Context) {
 	w := []func(){
 		func() {
-			pg.tableLayoutFunc(gtx, pg.theme.Body2("Transaction Fee"), pg.transactionFeeValueLabel)
+			pg.tableLayoutFunc(gtx, pg.theme.Body2("Transaction Fee"), pg.activeTransactionFeeValue, pg.inactiveTransactionFeeValue)
```
Comment (oshorefueled) [inline ui/send_page.go]: you don't need to add "Func" as a suffix to a method name.

File: ui/send_page.go [hunk 4]
```diff
@@ -560,44 +655,138 @@ func (pg *SendPage) validateAmount(ignoreEmpty bool) bool {
...
+func (pg *SendPage) getUSDValues(target interface{}) {
+	url := "https://api.bittrex.com/v3/markets/DCR-USDT/ticker"
+	resp, err := http.Get(url)
+	if err != nil {
+		return
+	}
+
+	defer resp.Body.Close()
+
+	json.NewDecoder(resp.Body).Decode(target)
```
Comment (oshorefueled) [inline ui/send_page.go]: This can be in the wallet package and add a comment about its function.

File: ui/send_page.go (hunk 2 above)
Comment (Sirmorrison) [reply ui/send_page.go]: The idea is to ensure the exchange is readily available at the point of typing the input and also not to make the app to freeze while trying to get the exchange data.

Comment (Sirmorrison) [reply ui/send_page.go]: Removing to go routine and moving it to the calculatedvalue function.

Comment (oshorefueled) [reply ui/send_page.go]: > The idea is to ensure the exchange is readily available at the point of typing the input and also not to make the app to freeze while trying to get the exchange data.

Still, calling an endpoint that frequent isn't an optimal solution to that. If real time time updates is what is needed, using sockets would be a better approach. Recursive calls to an endpoint, is a No No. Plus I don't think the exchange rates change that frequently to be called multiple times in a fraction of a second.

File: wallet/commands.go [hunk 5]
```diff
@@ -488,6 +490,19 @@ func (wal *Wallet) VerifyMessage(address string, message string, signature strin
...
+	resp, err := http.Get(url)
+	if err != nil {
+		return
+	}
+
+	defer resp.Body.Close()
+
+	json.NewDecoder(resp.Body).Decode(target)
+}
+
```
Comment (oshorefueled) [inline wallet/commands.go]: this doesn't belong here, it's not a wallet command. It should be in wallet.go just like the "GetBlockExplorerURL" method. Also, the comment should be more descriptive.   "GetUSDExchangeValues gets the exchange rate of DCR - USDT from a specified endpoint" would be more appropriate.

File: ui/send_page.go [hunk 6]
```diff
@@ -560,44 +649,127 @@ func (pg *SendPage) validateAmount(ignoreEmpty bool) bool {
...
+		txFeeValueUSD := dcrutil.Amount(txFee).ToCoin() * usdExchangeRate
+
+		totalAmountUSDTostring := fmt.Sprintf("%s USD", strconv.FormatFloat(totalAmountUSD, 'f', 7, 64))
+
+		pg.activeTotalAmount = dcrutil.Amount(amountAtoms).String()
+		pg.inactiveTotalAmount = totalAmountUSDTostring
+		pg.activeTransactionFeeValue = dcrutil.Amount(txFee).String()
+		pg.inactiveTransactionFeeValue = fmt.Sprintf("(%f USD)", txFeeValueUSD)
+		pg.activeTotalCostValue = dcrutil.Amount(totalCostDCR).String()
+		pg.inactiveTotalCostValue = fmt.Sprintf("(%s USD)", strconv.FormatFloat(totalAmountUSD+txFeeValueUSD, 'f', 7, 64))
```
Comment (oshorefueled) [inline ui/send_page.go]: This looks repeated. To make it cleaner, you can create a data type for it. For each case, assign the respective data properties their respective methods and pass that to a single method that initiates those values to the SendPage properties.
 Also, all that would be cleaner if they were in a separate method.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

File: ui/send_page.go [hunk 7]
```diff
@@ -560,44 +658,137 @@ func (pg *SendPage) validateAmount(ignoreEmpty bool) bool {
...
-	amountAtoms := int64(amount)
 
-	// set destination address
+	amountAtoms := int64(amount)
 	pg.txAuthor.RemoveSendDestination(0)
 	pg.txAuthor.AddSendDestination(pg.destinationAddressEditor.Text(), amountAtoms, false)
+	return amountAtoms
+}
+
+func (pg *SendPage) setUSDValues(usdExchangeRate, inputAmount float64, totalCostDCR, txFee, amountAtoms int64) {
```
Comment (oshorefueled) [inline ui/send_page.go]: make this a data type and pass a single argument to the Set method.

File: ui/send_page.go [hunk 8]
```diff
@@ -560,44 +658,137 @@ func (pg *SendPage) validateAmount(ignoreEmpty bool) bool {
...
+	// calculate total tx amount in DCR
+	txFeeValueUSD := dcrutil.Amount(txFee).ToCoin() * usdExchangeRate
+	totalAmountUSDTostring := fmt.Sprintf("%s USD", pg.sendAmountEditor.Text())
+
+	pg.activeTotalAmount = totalAmountUSDTostring
+	pg.inactiveTotalAmount = dcrutil.Amount(amountAtoms).String()
+	pg.activeTransactionFeeValue = fmt.Sprintf("%f USD", txFeeValueUSD)
+	pg.inactiveTransactionFeeValue = fmt.Sprintf("(%s)", dcrutil.Amount(txFee).String())
+	pg.activeTotalCostValue = fmt.Sprintf("%s USD", strconv.FormatFloat(inputAmount+txFeeValueUSD, 'f', 7, 64))
+	pg.inactiveTotalCostValue = fmt.Sprintf("(%s )", dcrutil.Amount(totalCostDCR).String())
```
Comment (oshorefueled) [inline ui/send_page.go]: you only need one method to set these.

File: ui/send_page.go [hunk 9]
```diff
@@ -560,44 +658,137 @@ func (pg *SendPage) validateAmount(ignoreEmpty bool) bool {
...
+	txFeeValueUSD := dcrutil.Amount(txFee).ToCoin() * usdExchangeRate
+
+	totalAmountUSDTostring := fmt.Sprintf("%s USD", strconv.FormatFloat(totalAmountUSD, 'f', 7, 64))
+
+	pg.activeTotalAmount = dcrutil.Amount(amountAtoms).String()
+	pg.inactiveTotalAmount = totalAmountUSDTostring
+	pg.activeTransactionFeeValue = dcrutil.Amount(txFee).String()
+	pg.inactiveTransactionFeeValue = fmt.Sprintf("(%f USD)", txFeeValueUSD)
+	pg.activeTotalCostValue = dcrutil.Amount(totalCostDCR).String()
+	pg.inactiveTotalCostValue = fmt.Sprintf("(%s USD)", strconv.FormatFloat(totalAmountUSD+txFeeValueUSD, 'f', 7, 64))
```
Comment (oshorefueled) [inline ui/send_page.go]: you only need one method to set these.

File: ui/send_page.go [hunk 10]
```diff
@@ -560,44 +658,137 @@ func (pg *SendPage) validateAmount(ignoreEmpty bool) bool {
...
+}
+
+func (pg *SendPage) updateDefaultValues(totalCostDCR, txFee, amountAtoms int64, noExchangeText string) {
+	pg.activeTotalAmount = dcrutil.Amount(amountAtoms).String()
+	pg.inactiveTotalAmount = noExchangeText
+	pg.activeTransactionFeeValue = dcrutil.Amount(txFee).String()
+	pg.inactiveTransactionFeeValue = ""
+	pg.activeTotalCostValue = dcrutil.Amount(totalCostDCR).String()
+	pg.inactiveTotalCostValue = ""
+}
```
Comment (oshorefueled) [inline ui/send_page.go]: this is the only method you need to set the data. And it only really needs to accept one argument.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

File: ui/send_page.go [hunk 11]
```diff
@@ -260,6 +309,8 @@ func (pg *SendPage) setAccountTabs() {
 }
 
 func (pg *SendPage) Layout(common pageCommon) {
+	pg.calculateValues()
```
Comment (oshorefueled) [inline ui/send_page.go]: this shouldn't be called on every frame.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

**Review by oshorefueled — APPROVED** [review]

**Pull request merged**